examples/arrowir/test_arrowir.cpp

```diff
@@ -0,0 +1,589 @@
+			break;
+		case duckdb::JoinType::RIGHT:
+			arrow_join_kind = arrowir::JoinKind_RightOuter;
+			break;
+		case duckdb::JoinType::ANTI:
+			arrow_join_kind = arrowir::JoinKind_Anti;
+			break;
+		default:
+			throw std::runtime_error("unsupported join type");
+		}
+
+		auto arrow_join_expression = transform_join_condition(fbb, join.conditions[0]);
+		auto function_name = fbb.CreateString("and");
+		for (auto i = begin(join.conditions) + 1, e = end(join.conditions); i != e; ++i) {
+			auto rhs_arrow = transform_join_condition(fbb, *i);
+			auto arg_list =
+			    fbb.CreateVector<flatbuffers::Offset<arrowir::Expression>>({arrow_join_expression, rhs_arrow});
+			auto call = arrowir::CreateCall(fbb, function_name, arg_list);
+			arrow_join_expression =
+			    arrowir::CreateExpression(fbb, arrowir::ExpressionImpl::ExpressionImpl_Call, call.Union());
+		}
+
+		auto arrow_join =
+		    arrowir::CreateJoin(fbb, rel_base, transform_op(fbb, *op.children[0]), transform_op(fbb, *op.children[1]),
+		                        arrow_join_expression, arrow_join_kind);
+		return arrowir::CreateRelation(fbb, arrowir::RelationImpl_Join, arrow_join.Union());
+	}
+
+	default:
+		throw std::runtime_error(duckdb::LogicalOperatorToString(op.type));
+	}
+}
+
+static duckdb::LogicalType transform_type_back(const arrow::Field *field) {
+	switch (field->type_type()) {
+	case arrow::Type_Binary:
+		return duckdb::LogicalType::VARCHAR;
+	case arrow::Type_Int: // TODO which size is this type? unclear.
+		return duckdb::LogicalType::BIGINT;
+	case arrow::Type_Date:
+		return duckdb::LogicalType::DATE;
+	case arrow::Type_Decimal:
+		// TODO need to transform those properly in transform_type
+		return duckdb::LogicalType::DECIMAL(10, 2);
+	default:
+		throw std::runtime_error(arrow::EnumNameType(field->type_type()));
+	}
+}
+
+static unique_ptr<duckdb::Expression> transform_expr_back(const arrowir::Expression *expr) {
+	switch (expr->impl_type()) {
+		//	case arrowir::ExpressionImpl::ExpressionImpl_FieldRef: {
+		//		return duckdb::make_unique<duckdb::BoundColumnRefExpression>();
+		//	}
+	default:
+		throw std::runtime_error(arrowir::EnumNameExpressionImpl(expr->impl_type()));
+	}
+}
+
+static unique_ptr<duckdb::LogicalOperator> transform_op_back(duckdb::ClientContext &context,
+                                                             const arrowir::Relation *arrow_rel) {
+	switch (arrow_rel->impl_type()) {
+	case arrowir::RelationImpl_OrderBy: {
+		auto arrow_op = arrow_rel->impl_as_OrderBy();
+		vector<duckdb::BoundOrderByNode> order_nodes;
+		// TODO fill nodes
+		auto res = duckdb::make_unique<duckdb::LogicalOrder>(move(order_nodes));
+		res->children.push_back(transform_op_back(context, arrow_op->rel()));
+		return res;
+	}
+	case arrowir::RelationImpl_Project: {
+		auto arrow_op = arrow_rel->impl_as_Project();
+		// TODO fill expressions
+		vector<unique_ptr<duckdb::Expression>> select_list;
+		// TODO where do we get our table index?
```

Comment:
Can you describe what this is? It might be missing from the Arrow IR, or called something else.

Reply:
Okay, after reading through the codebase, it looks like `table_index` is a global-to-`Binder` numbering of unique tables in a query. Is that correct? Arrow's IR doesn't currently have anything like that at the moment, but we can certainly see about adding that.

Reply:
In that case, would the numbering scheme be as follows:

```
tbl: 0
subquery: 1
outer query: 2
```

?

Reply:
Yes, but the numbering is on the relation level. One query might have multiple table_indexes. For example, the above query translates to the following plan:

```sql
D explain SELECT x+1 FROM (SELECT i+j FROM tbl) t(x);
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          +(x, 1)          │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          +(i, j)          │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            tbl            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             i             │
│             j             │
└───────────────────────────┘ 
```

Now if we add some aggregates, we get the following plan:

```sql
D explain SELECT x+1 FROM (SELECT SUM(i)+SUM(j) FROM tbl) t(x);

┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          +(x, 1)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│     +(sum(i), sum(j))     │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│      SIMPLE_AGGREGATE     │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           sum(i)          │
│           sum(j)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            tbl            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             i             │
│             j             │
└───────────────────────────┘
```

The aggregate refers to `tbl`, the first projection refers to the aggregate, and the second projection refers to the first projection, so they each have a unique table index.

Reply:
Thanks, got it. I'll open up a JIRA and PR to add this concept into the compute IR.

Reply:
JIRA: https://issues.apache.org/jira/browse/ARROW-14148
PR: https://github.com/apache/arrow/pull/11254/files